Minor optimizations on MooseVariableData

GiudGiud opened the pull request:
locally (M1 max) I observed 8% speedup on simple_diffusion.i with -r 7.
I am hoping someone can reproduce so we can merge this. 
I doubt the test suite will show a consistent improvement but let's see with this PR's run

GiudGiud committed:
Simplify a few loops in MooseVariableData
refs #29696

GiudGiud committed:
more memory locality and less if statements
refs #29696

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2640471/), step Docs: sync website on 94117f9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29697/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2640487/), step Generate coverage on 94117f9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b45164</th>
      <th colspan=4>#29697 94117f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.25%</td>
      <td>85.23%</td>
      <td>-0.02%</td>
      <td>65.58%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108039</td>
      <td>108022</td>
      <td>-17</td>
      <td>101</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18699</td>
      <td>18726</td>
      <td>+27</td>
      <td>53</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29697/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29697/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29697/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 65.58% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Avoid implicit conversions which prevent vectorization
And remove a few harmless others
Avoid a few size computations

GiudGiud committed:
Use same technique on computeAD for regular variables

GiudGiud committed:
Remove unused attributes

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/2651874/) on dd74a6c : invalidated by @loganharbour

MengnanLi91 commented:
I'll test this on my M1 and see how it goes

GiudGiud commented:
looking at the test suite, the Framework 1 and 2 recipe and Module
https://civet.inl.gov/branch/6695/

it seems we are running:
- Framework 1 5 minutes slower
- Framework 2 25 minutes faster
- Modules 12 minutes faster
but maybe it's just build machine differences

MengnanLi91 commented:
> looking at the test suite, the Framework 1 and 2 recipe and Module
> 
> https://civet.inl.gov/branch/6695/
> 
> 
> 
> it seems we are running:
> 
> - Framework 1 5 minutes slower
> 
> - Framework 2 25 minutes faster
> 
> - Modules 12 minutes faster
> 
> but maybe it's just build machine differences
> 
> 

I tried on my Mac M1 with mpiexec -n 10 -i simple_diffusion.i -r 7. Not much difference in total execution time

MengnanLi91 commented:
Not sure if the communication time  matters here. I can try a serial run tomorrow

GiudGiud commented:
dont use mpi, that will make the test times too small to look at

GiudGiud commented:
and I do nothing to improve MPI here

GiudGiud commented:
looking at more recipes, OpenMPI and ARM Mac, seems 0 change in runtime.

MengnanLi91 commented:
I ran it with perf_graph, the memory seems to change with new commit. I think it may have a direct effect from the changes

GiudGiud commented:
for the better or worse?

loganharbour committed:
Generalize filling to a lambda and combine RealEigenVector case

loganharbour committed:
Silence unused this

loganharbour commented:
Profiling reveals a factor of 2 improvement in `MooseVariableData::computeValues()` with simple diffusion, 7 refinements. This problem is the worst case for testing this as all we do is fetch vector tags. With problems that are more complex, the improvement will likely be better. Testing a transient problem now.

GiudGiud commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -424,537 +430,201 @@ MooseVariableData<OutputType>::computeValues()
-
-      if (is_transient)
+      if constexpr (is_real || is_real_vector)
+        dest[qp] = 0;
```
```suggestion
        dest[qp] = 0.;
```

GiudGiud replied:
Logan verified the cast was happening at compile time

GiudGiud commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -424,537 +430,201 @@ MooseVariableData<OutputType>::computeValues()
+    constexpr bool is_output = std::is_same_v<T, OutputType>;
+    constexpr bool is_gradient = std::is_same_v<T, OutputGradient>;
+    constexpr bool is_second = std::is_same_v<T, OutputSecond>;
+
```
should there be a ` is_div = is_same<T, lower rank(T) >` ?
the output type of div(vector) is going to be Real

GiudGiud replied:
```suggestion
    constexpr bool is_div = std::is_same_v<T, OutputDivergence>;
```

loganharbour committed:
Add fill for divergence

loganharbour committed:
Make assertions compile time errors, add comments

loganharbour committed:
Remove a branch

loganharbour committed:
Add an extra static assertion

loganharbour committed:
Correct static assertion

loganharbour committed:
Combine monomial optimization into the same function

loganharbour committed:
Fix ignores, remove unused captures

loganharbour committed:
Rename vector to dof_values

loganharbour committed:
Add extra assertion

loganharbour committed:
Expand always_false to allow multiple types, fix clang min bugs

loganharbour commented:
I did some verification about the inlining of the lambdas with our current GCC and clang with `-O0`: https://godbolt.org/z/94ra47jnK

GCC is smart enough to see that the version with the lambda is exactly the same as the one without, and actually just produces a jump.  Cool cool cool

loganharbour committed:
Only set starting with the 1st qp

GiudGiud reviewed (commented):
I approve

My only question would be whether the compiler will vectorize these lambdas (set_zero and accumulate).
At compile time they are exceedingly simple for each case so it's very likely. It's best to study this in a follow up

loganharbour commented:
Ping @lindsayad

lindsayad commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -1306,13 +826,13 @@ MooseVariableData<OutputType>::setDofValue(const OutputData & value, unsigned in
-    for (unsigned int i = 1; i < dof_values.size(); i++)
+  const auto nqps = u.size();
+  const auto ndofs = dof_values.size();
+  for (unsigned int qp = 0; qp < nqps; qp++)
+    u[qp] *= 0.;
+  for (unsigned int qp = 0; qp < nqps; qp++)
+    for (unsigned int i = 0; i < ndofs; i++)
```
```suggestion
  for (const auto qp : make_range(nqps))
    u[qp] *= 0.;
  for (const auto qp : make_range(nqps))
    for (const auto i : make_range(ndofs))
```

lindsayad commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -1326,12 +846,13 @@ MooseVariableData<OutputType>::setDofValues(const DenseVector<OutputData> & valu
-    for (unsigned int i = 1; i < dof_values.size(); i++)
+  const auto nqps = u.size();
+  const auto ndofs = dof_values.size();
+  for (unsigned int qp = 0; qp < nqps; qp++)
+    u[qp] *= 0.;
+  for (unsigned int qp = 0; qp < nqps; qp++)
+    for (unsigned int i = 0; i < ndofs; i++)
```
```suggestion
  for (const auto qp : make_range(nqps))
    u[qp] *= 0.;
  for (const auto qp : make_range(nqps))
    for (const auto i : make_range(ndofs))
```

lindsayad commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -1507,8 +1028,9 @@ MooseVariableData<RealEigenVector>::addSolution(NumericVector<Number> & sol,
-    for (unsigned int i = 0; i < _dof_indices.size(); ++i)
+    const auto num_dofs = _dof_indices.size();
+    unsigned int inc = (isNodal() ? j : j * num_dofs);
+    for (unsigned int i = 0; i < num_dofs; ++i)
```
```suggestion
    for (const auto i : make_range(num_dofs))
```

lindsayad commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -1599,7 +1121,7 @@ MooseVariableData<OutputType>::computeIncrementAtQps(const NumericVector<Number>
   {
-    _increment[qp] = 0;
+    _increment[qp] = 0.;
     for (unsigned int i = 0; i < num_dofs; i++)
```
```suggestion
    for (const auto i : make_range(num_dofs))
```

lindsayad commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -1669,10 +1191,11 @@ MooseVariableData<RealEigenVector>::computeIncrementAtNode(
   {
     unsigned int n = 0;
+    const auto n_dof_indices = _dof_indices.size();
     for (unsigned int j = 0; j < _count; j++)
```
```suggestion
    for (const auto j : make_range(_count))
```

GiudGiud commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -416,545 +422,243 @@ MooseVariableData<OutputType>::divPhiFace() const
-      if (_need_vector_tag_grad[tag])
-        _vector_tag_grad[tag][i] = 0;
+      _mapped_grad_phi.resize(num_dofs);
+      for (unsigned int i = 0; i < num_dofs; i++)
```
We ll need to check that these vectorize as well

GiudGiud replied:
reading up on it it seems they compile to the same code. I ll make all the changes

GiudGiud committed:
Address Alex's review

- use range based loops for cleaner code
- use auto a little more

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
Valgrind subchannel failure is because subchannel is not merged yet. Unrelated

GiudGiud commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -416,545 +422,243 @@ MooseVariableData<OutputType>::divPhiFace() const
-        // Note: this does NOT do any allocation.  It is "reconstructing" the object in place
-        new (&_mapped_grad_phi_face[i][qp])
-            Eigen::Map<RealDIMValue>(const_cast<Real *>(&(*_current_grad_phi)[i][qp](0)));
+      for (unsigned int qp = 0; qp < nqp; ++qp)
+        set_zero(qp);
+      for (unsigned int i = 0; i < num_dofs; i++)
+        for (unsigned int qp = 0; qp < nqp; qp++)
```
will catch later

GiudGiud commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -416,545 +422,243 @@ MooseVariableData<OutputType>::divPhiFace() const
+    {
       _du_dot_du.resize(nqp);
-
+      for (unsigned int i = 0; i < nqp; ++i)
+        _du_dot_du[i] = 0.;
+      for (unsigned int i = 0; i < num_dofs; i++)
+        for (unsigned int qp = 0; qp < nqp; qp++)
+          _du_dot_du[qp] = _dof_du_dot_du[i];
+    }
     if (_need_du_dotdot_du)
+    {
       _du_dotdot_du.resize(nqp);
+      for (unsigned int i = 0; i < nqp; ++i)
+        _du_dotdot_du[i] = 0.;
+      for (unsigned int i = 0; i < num_dofs; i++)
+        for (unsigned int qp = 0; qp < nqp; qp++)
```
will catch later

GiudGiud committed:
Use make_range in more loops in same class
Avoid capturing index range by reference in case that triggers an undesired lookup

GiudGiud commented:
had missed 3 of your comments. they are integrated now, along with a couple more make_range. Minor changes, no need for a new approval imo
